src/main/java/org/manifold/compiler/back/digital/SMT2CodeGenerator.java

```diff
@@ -0,0 +1,299 @@
+    PortValue inputPort = node.getPort("out");
+    Net inputNet = netlist.getConnectedNet(inputPort);
+    String netName = inputNet.getName();
+    for (int i = 0; i <= numberOfStates; ++i) {
+      Symbol symInput = QFABV.getStateVariable(nodeName, i);
+      Symbol symNet = QFABV.getStateVariable(netName, i);
+      // declare state variables
+      declarations.add(QFABV.declareBitVector(symInput, 1));
+      declarations.add(QFABV.declareBitVector(symNet, 1));
+      // the input drives the net on every timestep
+      assertions.add(QFABV.assertThat(QFABV.equal(symInput, symNet)));
+    }
+  }
+  
+  private void generateOutputPin(String nodeName, NodeValue node) 
+      throws UndeclaredIdentifierException {
+    // find out what net drives the input
+    PortValue outputPort = node.getPort("in");
+    Net outputNet = netlist.getConnectedNet(outputPort);
+    String netName = outputNet.getName();
+    for (int i = 0; i <= numberOfStates; ++i) {
+      Symbol symOutput = QFABV.getStateVariable(nodeName, i);
+      Symbol symNet = QFABV.getStateVariable(netName, i);
+      // declare state variables (outputs only)
+      declarations.add(QFABV.declareBitVector(symOutput, 1));
+      // the input drives the net on every timestep
+      assertions.add(QFABV.assertThat(QFABV.equal(symOutput, symNet)));
+    }
+  }
+  
+  private void generateAndGate(String nodeName, NodeValue node) 
+      throws UndeclaredIdentifierException {
+    // out = in0 AND in1
+    PortValue in0Port = node.getPort("in0");
+    PortValue in1Port = node.getPort("in1");
+    PortValue outPort = node.getPort("out");
+    
+    Net in0Net = netlist.getConnectedNet(in0Port);
+    Net in1Net = netlist.getConnectedNet(in1Port);
+    Net outNet = netlist.getConnectedNet(outPort);
+    
+    String in0NetName = in0Net.getName();
+    String in1NetName = in1Net.getName();
+    String outNetName = outNet.getName();
+    
+    for (int i = 0; i <= numberOfStates; ++i) {
+      Symbol symIn0 = QFABV.getStateVariable(in0NetName, i);
+      Symbol symIn1 = QFABV.getStateVariable(in1NetName, i);
+      Symbol symOut = QFABV.getStateVariable(outNetName, i);
+      // declare net for output
+      declarations.add(QFABV.declareBitVector(symOut, 1));
+      assertions.add(QFABV.assertThat(QFABV.equal(
+          symOut, QFABV.and(symIn0, symIn1))));
+    }
+  }
+  
+  private void generateOrGate(String nodeName, NodeValue node) 
+      throws UndeclaredIdentifierException {
+    // out = in0 OR in1
+    PortValue in0Port = node.getPort("in0");
+    PortValue in1Port = node.getPort("in1");
+    PortValue outPort = node.getPort("out");
+    
+    Net in0Net = netlist.getConnectedNet(in0Port);
+    Net in1Net = netlist.getConnectedNet(in1Port);
+    Net outNet = netlist.getConnectedNet(outPort);
+    
+    String in0NetName = in0Net.getName();
+    String in1NetName = in1Net.getName();
+    String outNetName = outNet.getName();
+    
+    for (int i = 0; i <= numberOfStates; ++i) {
```

Comment:
I feel like a bunch of common code here can be shared as well. Seems that the only difference is ```QFABV.and``` vs ```QFABV.or```, which I believe can be passed in like so:

```private void generateGate(BiConsumer<Symbol, Symbol>, String nodeName, NodeValue node)
...
assertions.add(QFABV.assertThat(QFABV.equal(symOut, gate.accept(symIn0, symIn1))```

and just call it with ```generateGate(QFABV::and, ..., ...);```